browser/web_discovery/wdp_service_factory.cc

```diff
@@ -0,0 +1,63 @@
+/* Copyright (c) 2024 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#include "brave/browser/web_discovery/wdp_service_factory.h"
+
+#include "base/path_service.h"
+#include "brave/components/web_discovery/browser/wdp_service.h"
+#include "brave/components/web_discovery/common/features.h"
+#include "chrome/browser/browser_process.h"
+#include "chrome/common/chrome_paths.h"
+#include "components/keyed_service/content/browser_context_dependency_manager.h"
+#include "components/user_prefs/user_prefs.h"
+#include "content/public/browser/browser_context.h"
+#include "content/public/browser/storage_partition.h"
+
+namespace web_discovery {
+
+WDPService* WDPServiceFactory::GetForBrowserContext(
+    content::BrowserContext* context) {
+  return static_cast<WDPService*>(
+      GetInstance()->GetServiceForBrowserContext(context, true));
+}
+
+WDPServiceFactory* WDPServiceFactory::GetInstance() {
+  static base::NoDestructor<WDPServiceFactory> instance;
+  return instance.get();
+}
+
+WDPServiceFactory::WDPServiceFactory()
+    : BrowserContextKeyedServiceFactory(
+          "WDPService",
+          BrowserContextDependencyManager::GetInstance()) {}
+
+WDPServiceFactory::~WDPServiceFactory() = default;
+
+KeyedService* WDPServiceFactory::BuildServiceInstanceFor(
+    content::BrowserContext* context) const {
+  auto* default_storage_partition = context->GetDefaultStoragePartition();
+  auto shared_url_loader_factory =
+      default_storage_partition->GetURLLoaderFactoryForBrowserProcess();
+  base::FilePath user_data_dir =
+      base::PathService::CheckedGet(chrome::DIR_USER_DATA);
+  return new WDPService(g_browser_process->local_state(),
+                        user_prefs::UserPrefs::Get(context), user_data_dir,
+                        shared_url_loader_factory);
+}
+
+content::BrowserContext* WDPServiceFactory::GetBrowserContextToUse(
+    content::BrowserContext* context) const {
+  if (!base::FeatureList::IsEnabled(features::kWebDiscoveryNative)) {
+    return nullptr;
+  }
+  // Prevents creation of service instance for incognito/OTR profiles
+  return context->IsOffTheRecord() ? nullptr : context;
```

Comment:
can we add a test to make sure it's not running in private windows?

Reply:
added in 4ec17ed3ddf4ed1cef849e88b1e40c46c7563c45